### PR TITLE
Add OMP support for ARCH host backend

### PR DESCRIPTION
**Pull request by hokkanen**
This PR adds OpenMP reduction support for arch host backend. OpenMP parallelization is handled completely in the arch header and no parallel region is needed in datareductionoperator.cpp. Brings datareducers back to about same speed as in dev branch (huge speedup).

### CODE REVIEW

**Comment by markusbattarbee** [comment]
Ok, so if I read this correct, the only change from this to when running on GPU is that it won't needlessly create and then delete an OpenMP parallel region for each cell?

**Comment by markusbattarbee** [comment]
Nevertheless, it is indeed a lot faster when running on GPU:
```
(MAIN): All timesteps calculated.
         (TIME) total run time 1340.6 s, total simulated time 30.2744 s
         (TIME) seconds per timestep 17.8747, seconds per simulated second 44.2817
```
versus
```
(MAIN): All timesteps calculated.
         (TIME) total run time 5037.4 s, total simulated time 30.2744 s
         (TIME) seconds per timestep 67.1654, seconds per simulated second 166.391
```
👍

File: datareduction/datareductionoperator.cpp [hunk 1]
```diff
@@ -1862,7 +1862,7 @@ namespace DRO {
 
    bool VariableHeatFluxVector::reduceData(const SpatialCell* cell,char* buffer) {
       const Real HALF = 0.5;
-      # pragma omp parallel
+      #pragma omp parallel
```
Comment (markusbattarbee) [inline datareduction/datareductionoperator.cpp:1865]: I guess this new heat flux also needs porting to ARCH?

File: datareduction/datareductionoperator.cpp [hunk 2]
```diff
@@ -1642,7 +1642,7 @@ namespace DRO {
          }
       }
 
-      # pragma omp parallel
+      #pragma omp parallel
```
Comment (markusbattarbee) [inline datareduction/datareductionoperator.cpp:1645]: Can this  VariablePrecipitationLineDiffFlux also be ported to ARCH?

File: datareduction/datareductionoperator.cpp [hunk 3]
```diff
@@ -2015,7 +2015,7 @@ namespace DRO {
    // Accumulate contributions coming from this velocity block to the
    // spatial cell velocity moments. If multithreading / OpenMP is used,
    // these updates need to be atomic:
-#pragma omp critical
+///#pragma omp critical
          { epsilon += thread_epsilon_sum; }
```
Comment (markusbattarbee) [inline datareduction/datareductionoperator.cpp:2019]: Opinions @ursg @ykempf - do we want to remove the commented out omp pragmas from this file? And the critical region curlies? If they are all going to be handled by ARCH anyway.

Comment (ursg) [reply datareduction/datareductionoperator.cpp:2019]: In principle, yes. If it works, and guarantees consistency and all.
Should we leave a friendly warning comment there though? For the future?

**Comment by hokkanen** [comment]
> Ok, so if I read this correct, the only change from this to when running on GPU is that it won't needlessly create and then delete an OpenMP parallel region for each cell?

Yes, influences GPU exec as well.

File: datareduction/datareductionoperator.cpp (hunk 2 above)
Comment (hokkanen) [reply datareduction/datareductionoperator.cpp]: > Can this VariablePrecipitationLineDiffFlux also be ported to ARCH?

I added arch for all three remaining loops now.

File: datareduction/datareductionoperator.cpp (hunk 1 above)
Comment (hokkanen) [reply datareduction/datareductionoperator.cpp]: Done.

File: datareduction/datareductionoperator.cpp (hunk 3 above)
Comment (hokkanen) [reply datareduction/datareductionoperator.cpp:2021]: I didn't do this yet, maybe if you have an idea about what kind of warnings to keep, please feel free to add commits (commits allowed for maintainers checked).

**Comment by markusbattarbee** [comment]
Looks like there's some bug in one of the last three additions - Ionosphere_small (which includes all reducers) now crashes during testpackage:
```
running Ionosphere_small 
We detected OpenMPI so we set the cvars value to disable ompio, MCA io: ^ompio
Note: Initializing Magnetosphere with dipole type 4, enforcing no dipole in solar wind!
(Grid) rank 0 is noderank 0 of 1
Done setting all 64 instances of device mesh wrapper handler!
Maximum refinement level is 1
Finished first level of refinement
(MAIN): Completed grid initialization.
(MAIN): Starting main simulation loop.


an illegal memory access was encountered in datareduction/../arch/arch_device_cuda.h at line 479


driver shutting down in spatial_cell_gpu.hpp at line 277


driver shutting down in spatial_cell_gpu.hpp at line 277
srun: error: g5102: task 0: Exited with exit code 1
srun: Terminating StepId=3476662.47

**Comment by hokkanen** [comment]
> Looks like there's some bug in one of the last three additions - Ionosphere_small (which includes all reducers) now crashes during testpackage, at the first file output:
> 
> ```
> running Ionosphere_small 
> We detected OpenMPI so we set the cvars value to disable ompio, MCA io: ^ompio
> Note: Initializing Magnetosphere with dipole type 4, enforcing no dipole in solar wind!
> (Grid) rank 0 is noderank 0 of 1
> Done setting all 64 instances of device mesh wrapper handler!
> Maximum refinement level is 1
> Finished first level of refinement
> (MAIN): Completed grid initialization.
> (MAIN): Starting main simulation loop.
> 
> 
> an illegal memory access was encountered in datareduction/../arch/arch_device_cuda.h at line 479
> 
> 
> driver shutting down in spatial_cell_gpu.hpp at line 277
> 
> 
> driver shutting down in spatial_cell_gpu.hpp at line 277
> srun: error: g5102: task 0: Exited with exit code 1
> srun: Terminating StepId=3476662.47
> ```
> 
> Not sure which reducer it is, that requires running them one by one and won't have time for that today.

This issue is now fixed:
```
running Ionosphere_small 
We detected OpenMPI so we set the cvars value to disable ompio, MCA io: ^ompio
Done setting all 64 instances of device mesh wrapper handler!
Maximum refinement level is 1
Finished first level of refinement
(MAIN): Completed grid initialization.
(MAIN): Starting main simulation loop.
(MAIN): Completed requested simulation. Exiting.
--------------------------------------------------------------------------------------------
Ionosphere_small  -  Verifying 4ec8851e5bcd02802f41216e8745b010ba0c9862___DACC_SEMILAG_PQM__DTRANS_SEMILAG_PPM__DDP__DDPF__DVEC_FALLBACK_GENERIC__DVECL=16 against 959a6ac229670b29cd43c6fce3108e538ded098f__DACC_SEMILAG_PQM__DTRANS_SEMILAG_PPM__DDP__DDPF__DVEC_FALLBACK_GENERIC__DVECL
--------------------------------------------------------------------------------------------
------------------------------------------------------------
 ref-time     |   new-time       |  speedup                |
------------------------------------------------------------
6569        1218         5.39327
------------------------------------------------------------
  variable     |     absolute diff     |     relative diff | 
------------------------------------------------------------
Comparing file /scratch/project_2010357/jarohokk/vlas/vlasiator/testpackage/run_2024.06.03_11.22.43/Ionosphere_small/bulk.0000001.vlsv against reference
proton/vg_rho_0                8.91876e-05                 3.86797e-11
proton/vg_v_0                7.76618e-06                 1.03523e-11
proton/vg_v_1                2.08696e-06                 3.93569e-10
proton/vg_v_2                1.74844e-05                 6.03424e-09
fg_b_0                2.05339e-20                 2.07029e-13
fg_b_1                6.1261e-21                 6.17525e-14
fg_b_2                1.39529e-20                 2.04643e-16
fg_e_0                1.05778e-14                 2.40125e-10
fg_e_1                6.22805e-14                 1.56708e-11
fg_e_2                6.6584e-15                 1.91168e-11
ig_upmappedarea_0                42.1406                 5.10665e-13
ig_fac_0                7.99643e-19                 6.87365e-11
ig_rhon_0                4.68642e-06                 2.3357e-12
ig_potential_0                4.14386e-14                 7.63512e-11
VLSV file timestamps match.
--------------------------------------------------------------------------------------------
Comparing file /scratch/project_2010357/jarohokk/vlas/vlasiator/testpackage/run_2024.06.03_11.22.43/Ionosphere_small/bulk.0000002.vlsv against reference
proton/vg_rho_0                7.43981e-05                 2.70289e-11
proton/vg_v_0                8.20402e-06                 1.09334e-11
proton/vg_v_1                6.43295e-06                 4.05452e-10
proton/vg_v_2                1.52085e-05                 2.66315e-09
fg_b_0                6.6578e-20                 6.71259e-13
fg_b_1                1.41536e-20                 1.42619e-13
fg_b_2                4.35891e-20                 6.3931e-16
fg_e_0                1.49513e-14                 1.15526e-10
fg_e_1                3.50679e-14                 8.81571e-12
fg_e_2                1.21553e-14                 3.42544e-11
ig_upmappedarea_0                128.094                 1.5664e-12
ig_fac_0                2.11072e-18                 7.19151e-11
ig_rhon_0                1.12087e-05                 5.24912e-12
ig_potential_0                9.6981e-14                 9.67644e-11
VLSV file timestamps match.
--------------------------------------------------------------------------------------------
--------------------------------------------------------------------------------------------

```

**Pull request merged**